### PR TITLE
Add schema for stories, together with tests

### PR DESCRIPTION
**Pull request by thrix**

### CODE REVIEW

**Comment by thrix** [comment]
Based on #1202

**Comment by happz** [comment]
With stories using many keys from the base set, this calls for extracting them to some `base.yaml` schema, and re-use them in both `tests.yaml` and `stories.yaml` and `plans.yaml` and so on. And I think we should define not just complex objects like fmf id or `adjust`, but even the most basic keys like `description`, because we do need them to be the same in all nodes. Even if they are just the simplest possible primitive type like `string`, all levels pointing to the very same definition make it clear and easier to maintain in case of changes to what these keys may contain.

```yaml
# base.yaml
definitions:
  description:
    type: string

# tests.yaml
...
properties:
  description:
    "$ref": "base.yaml#/definitions/description"

# plans.yaml
...
properties:
  description:
    "$ref": "base.yaml#/definitions/description"

# stories.yaml
...
properties:
  description:
    "$ref": "base.yaml#/definitions/description"
```

**Comment by thrix** [comment]
@happz yep, of course. So, how to appoach it? Let fmf take multiple schemas and create a store from it, I guess should be that easy right?

**Comment by happz** [comment]
As discussed, "schema store" seems like the expected way to go. For maintenance purposes, it would be better to separate schemas according to "tree levels" similar to TMT specifications (Core, Test, Plan, ...), with Core being shared as a "library" to serve other levels. And "store" seems like the way to go with JSON schema validator to be able to resolve all those `$ref` links.

**Comment by ukulekek** [comment]
LGTM

**Review by ukulekek — APPROVED** [review]

**Comment by thrix** [comment]
but good for review

File: tmt/schemas/stories.yaml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  # https://tmt.readthedocs.io/en/stable/spec/core.html#link
+  link:
+    $ref: "/schemas/core#/definitions/link"
+
+  # https://tmt.readthedocs.io/en/stable/spec/core.html#order
+  order:
+    $ref: "/schemas/core#/definitions/order"
+
+  # https://tmt.readthedocs.io/en/stable/spec/stories.html#priority
+  # TODO: missing from documentation
```
Comment (happz) [inline tmt/schemas/stories.yaml:42]: FTR, https://github.com/teemtee/tmt/pull/440 adds the specification.

Comment (thrix) [reply tmt/schemas/stories.yaml:42]: Unfortunately, #440 will need more discussion, the meeting is scheduled for Thursday, I would say we will need to update this once we are clear on #440.

Comment (thrix) [reply tmt/schemas/stories.yaml:42]:
```suggestion
# TODO: might need update once https://github.com/teemtee/tmt/pull/440 is merged
```

File: tmt/schemas/stories.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  # https://tmt.readthedocs.io/en/stable/spec/core.html#order
+  order:
+    $ref: "/schemas/core#/definitions/order"
+
+  # https://tmt.readthedocs.io/en/stable/spec/stories.html#priority
+  # TODO: might need update once https://github.com/teemtee/tmt/pull/440 is merged
+  priority:
+    type: string
+    enum:
+      - should
```
Comment (thrix) [inline tmt/schemas/stories.yaml:46]:
```suggestion
      - could
      - should
```

**Comment by thrix** [comment]
I believe this is good to go now

**Review by happz — APPROVED** [review]

**Review by guy9050 — APPROVED** [review]

File: spec/tests/framework.fmf [hunk 3]
```diff
@@ -27,7 +27,7 @@ description: |
         Exit code and BeakerLib's ``TestResults`` file determine
         the test result.
 
-example:
+example: |
```
Comment (psss) [inline spec/tests/framework.fmf:30]: This are two examples, they should be formatted as two separate sections.

File: spec/tests/path.fmf [hunk 4]
```diff
@@ -10,7 +10,7 @@ description: |
     the test script, automation changes the current working
     directory to the provided ``path`` before running the test.
 
-    It should be a ``string`` containing path from the metadata
+    It must be a ``string`` containing path from the metadata
```
Comment (psss) [inline spec/tests/path.fmf:13]: Let's handle this in a separate PR as agreed in the `test` spec.

File: tests/unit/test_schemas.py [hunk 5]
```diff
@@ -15,35 +15,64 @@
...
-    schema = YAML(typ="safe").load(open(schema_file, encoding="utf-8"))
+    return YAML(typ="safe").load(open(schema_file, encoding="utf-8"))
 
-    schema_store = {}
+
+@pytest.fixture
+def stories_schema():
+    # TODO: tmt package shall provide a helper function
+    # for "load schemas for test/plan/story"
+    schema_file = os.path.join(SCHEMADIR, 'stories.yaml')
```
Comment (psss) [inline tests/unit/test_schemas.py:29]: Let's rename to `story.yaml` as @happz proposed.

File: tmt/schemas/stories.yaml [hunk 6]
```diff
@@ -0,0 +1,63 @@
+---
+
+#
+# JSON Schema definition for tmt `Stories`
+#
+# https://tmt.readthedocs.io/en/stable/spec/stories.html
+#
+
+$id: /schemas/stories
```
Comment (psss) [inline tmt/schemas/stories.yaml:9]:
```suggestion
$id: /schemas/story
```

I guess the id should be updated as well, right?

Comment (thrix) [reply tmt/schemas/stories.yaml:9]: yeah

File: spec/tests/framework.fmf (hunk 3 above)
Comment (thrix) [reply spec/tests/framework.fmf:30]: @psss how, can you suggest?

File: tests/unit/test_schemas.py (hunk 5 above)
Comment (thrix) [reply tests/unit/test_schemas.py:29]: renamed in next update

**Review by psss — APPROVED** [review]
Thanks for all the improvements! Looks good. Just fixed the example, squashed and rebased.

**Pull request merged**